okhttp/src/jvmMain/kotlin/okhttp3/Dispatcher.kt

```diff
@@ -86,41 +86,49 @@ class Dispatcher() {
   @get:Synchronized
   var idleCallback: Runnable? = null
 
-  private var executorServiceOrNull: ExecutorService? = null
-
   @get:Synchronized
-  @get:JvmName("executorService") val executorService: ExecutorService
-    get() {
-      if (executorServiceOrNull == null) {
-        executorServiceOrNull = ThreadPoolExecutor(0, Int.MAX_VALUE, 60, TimeUnit.SECONDS,
-            SynchronousQueue(), threadFactory("$okHttpName Dispatcher", false))
-      }
-      return executorServiceOrNull!!
-    }
+  @set:Synchronized
+  @get:JvmName("priorityOrder")
+    /**
+     * A comparator that determines the priority of the async calls.
+     * A call with the highest priority will be executed first.
+     *
+     * When [priorityOrder] is null or returns 0, calls that are enqueued first have priority.
+     */
+  var priorityOrder: Comparator<Call>? = null
 
   /** Ready async calls in the order they'll be run. */
-  private val readyAsyncCalls = ArrayDeque<AsyncCall>()
+  private val readyAsyncCalls = LinkedList<AsyncCall>()
```

Comment:
I wonder whether this might be simpler with a very small DispatchQueue model.

say

```
interface DispatchQueue {
  fun insert(call: AsyncCall)
  fun iterator(): Iterator<AsyncCall>
  fun sort()
}

StandardDispatchQueue: DispatchQueue {
  private val calls = ArrayDeque<AsyncCall>()

  fun insert(call: AsyncCall) {
    calls.add(call)
  }

  fun iterator(): Iterator<AsyncCall> {
    return calls.iterator()
  }

  fun sort() {
    // Nothing
  }
}

PriorityDispatchQueue(private val comparator = Comparator<Call>): DispatchQueue {
  private var calls = LinkedList<AsyncCall>()

  fun insert(call: AsyncCall) {
    val pos = l.binarySearch(call, comparator)
    if (pos >= 0) {
      // walk forward to find position
      l.add(finalPos, call)
    } else {
      l.add(-pos - 1, call)
    }
  }

  fun iterator(): Iterator<AsyncCall> {
    return calls.iterator()
  }

  fun sort() {
    calls.sort(comparator)
  }
}
```

Two reasons

1) We can remain reasonably confident that the existing behaviour remains, and is no more expensive.
2) We can optimise and test this code separately, and avoid walking the entire list if big.  If small, it's hopefully as efficient. Currently the Dispatcher assumes all the new complexity, it would be nice to leave it as simple or simpler than before.

Thoughts? Code above is rough and probably doesn't even compile.

Reply:
@yschimke I think you can't binarySearch over LinkedList, only linear search works. Even if you copy the whole list to array and find the index, you need to traverse to that index anyways.

Reply:
Maybe I am not fully understanding how much this code should be able to handle. What is the worst case scenario this code has to handle? Until now I assumed maybe 1000 incoming calls and 100 maxRequests.

Reply:
> y understanding how much this code should be able to handle. What is the worst case scenario this code has to handle? Until now I assumed maybe 1000 incoming ca

Good point.  I just failed this interview :)

Maybe that's a good microbenchmark.  submit a 1000 requests and see how often comparator is called?

Reply:
Your usage, my usage doesn't necessarily represent all usage. I'd be concerned about apps using this with huge workloads and it spending any significant time here.  Also why I'm nervous about more open design like `nextCall(allCalls)`.

Reply:
> Also why I'm nervous about more open design like nextCall(allCalls).

yes, I totally understand you. I just wanted to know what use case you are considering and align our thoughts.